### ISSUE TITLE
How to deal with the "mode collapse" problem

### ISSUE THREAD

**Issue by RefDawn-XD** [feiglab/idpgan#3]
Hi, when I tried to follow your work, I encountered a problem called mode collapse. This means that the generator could only generate several similar conformations, but lacked the ability to generate multiple diverse conformations. Have you encountered the same problem when training your model?

**Comment by giacomo-janson** [comment]
Dear lizongquan01,

yes, we encountered the same problem when training a GAN model. Unfortunately, this happens all the time when using the original GAN formulation. Also training is highly unstable.
To avoid mode collapse and stabilize training, we had to use modified GAN versions. Did you try this? The two versions that worked well for us are Spectral Normalization GANs (paper: https://arxiv.org/abs/1802.05957) and Wasserstein GANs with gradient penalty (paper: https://arxiv.org/abs/1704.00028). Probably SN-GANs are easier to implement and faster to train and that's what we used mostly. Also, the Top-k training technique (paper: https://arxiv.org/abs/2002.06224) is simple and appears to stabilize all GAN version in my experience, including vanilla GANs sometimes. The idpGAN paper reports the choices we made for the final models.
Finally, also consider that more recent types of generative models are out there nowadays. Diffusion models, and recently flow matching models, are way more stable than GANs and typically beat GAN performance. Everyone seems to be using them right now.
Best

Giacomo

**Comment by RefDawn-XD** [comment]
Thank you for your kindly detailed reply! I have tried Spectral Normalization methods, but still meet the unstable training problem. I will try to modify the loss and use tok-k technique later in order to get a stable version. 
Sure, It would be exciting to see these models if they can put to good use  : )

**Comment by giacomo-janson** [comment]
Yes, unfortunately training GANs is mostly a trial-and-error process.
Btw, there sure are many interesting applications for these kind of generative models even right now!
